npm install fails when node_modules is a symlink

Steps to reproduce:

```
cd `mktemp -d`
npm init --force
ln -s `mktemp -d` node_modules
npm install --save-dev typescript@1.7.0-dev.20151006
npm install --save-dev typescript@1.7.0-dev.20151014
```

Gives the following warning:

```
npm WARN update-linked node_modules/typescript needs updating to 1.7.0-dev.20151014 from 1.7.0-dev.20151006 but we can't, as it's a symlink
```

which should be an error because the install has **not** updated the installed version:

```
$ ./node_modules/.bin/tsc --version
message TS6029: Version 1.7.0-dev.20151006
```

Repeat the above but without a symlinked `node_modules` and it works just fine:

```
$ ./node_modules/.bin/tsc --version
message TS6029: Version 1.7.0-dev.20151014
```

Important details:

```
$ node --version
v4.2.0
$ npm --version
3.3.8
```

I personally didn't have a problem with `npm install` and `node_modules` being a symlink, however when trying to run `npm run build` for the build process in the local package, I get the following error:
`Cannot find module 'gulp-util'`

My guess is it simply can't resolve the path of gulp-util, which is probably the first requested child-dependency, as my build script calls gulp.

This was encountered this while trying to fix long path names on a Windows hosted VM with a shared folder. The symlink was used to stop windows inheriting the files inside the share, and instead divert them to somewhere else on the VM.  This was then reproduced on the raw VM outside of any share paths, using `ubuntu/trusty64`.

Steps to reproduce, you can probably try this with any project that has sub-deps;
- `npm install`
- `npm run build`

Should complete OK
- `mv node_modules /tmp/nm`
- `ln -s /tmp/nm node_modules`
- `npm run build`

Observe failure.

npm 3.3.12, node 5.4.1

After upgrade to npm 3.x we faced similar problem in our CI: we have a symlink to `node_modules` and launch `npm prune` and `npm update`. It causes **No compatible version found: MODULE_NAME@linked** error.

Is there any workaround besides downgrade to npm 2.x?

npm output:
```
$ npm update
npm ERR! Linux 3.10.0-123.el7.x86_64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "update"
npm ERR! node v5.8.0
npm ERR! npm  v3.7.3

npm ERR! No compatible version found: lodash@linked
npm ERR! Valid install targets:
npm ERR! 4.6.1, 4.6.0, 4.5.1, 4.5.0, 4.4.0, 4.3.0, 4.2.1, 4.2.0, 4.1.0, 4.0.1, 4.0.0, 3.10.1, 3.10.0, 3.9.3, 3.9.2, 3.9.1, 3.9.0, 3.8.0, 3.7.0, 3.6.0, 3.5.0, 3.4.0, 3.3.1, 3.3.0, 3.2.0, 3.1.0, 3.0.1, 3.0.0, 2.4.2, 2.4.1, 2.4.0, 2.3.0, 2.2.1, 2.2.0, 2.1.0, 2.0.0, 1.3.1, 1.3.0, 1.2.1, 1.2.0, 1.1.1, 1.1.0, 1.0.2, 1.0.1, 1.0.0, 1.0.0-rc.3, 1.0.0-rc.2, 1.0.0-rc.1, 0.10.0, 0.9.2, 0.9.1, 0.9.0, 0.8.2, 0.8.1, 0.8.0, 0.7.0, 0.6.1, 0.6.0, 0.5.2, 0.5.1, 0.5.0, 0.5.0-rc.1, 0.4.2, 0.4.1, 0.4.0, 0.3.2, 0.3.1, 0.3.0, 0.2.2, 0.2.1, 0.2.0, 0.1.0
npm ERR!
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /sdnc_users/adj/aimsys/sdnc01/npm-update/2/npm-debug.log
```

--- a/package.json
+++ b/package.json
@@ -15,6 +15,7 @@
   },
   "devDependencies": {
     "archy": "0",
+    "tacks": "^1.2.1",
     "tap": "^1.2.0"
   },
   "scripts": {

--- a/rpt.js
+++ b/rpt.js
@@ -125,7 +125,14 @@ function loadChildren (node, cache, filterWith, cb) {
   // (or more) time.
   cb = once(cb)
   var nm = path.join(node.path, 'node_modules')
-  return readdir(nm, thenLoadKids)
+  var rm
+  return fs.realpath(path.join(node.path, 'node_modules'), thenReaddir)
+
+  function thenReaddir (er, real_nm) {
+    if (er) return cb(null, node)
+    rm = real_nm
+    readdir(nm, thenLoadKids)
+  }
 
   function thenLoadKids (er, kids) {
     // If there are no children, that's fine, just return
@@ -139,7 +146,7 @@ function loadChildren (node, cache, filterWith, cb) {
   }
   function thenLoadNode (kid, done) {
     var kidPath = path.join(nm, kid)
-    var kidRealPath = path.join(node.realpath,'node_modules',kid)
+    var kidRealPath = path.join(rm, kid)
     loadNode(kidPath, kidRealPath, cache, andAddNode(done))
   }
   function andAddNode (done) {

--- a/rpt.js
+++ b/rpt.js
@@ -80,7 +80,7 @@ function loadNode (logical, physical, cache, cb) {
   fs.realpath(physical, function (er, real) {
     if (er) return cb(er)
     debug('realpath l=%j p=%j real=%j', dpath(logical), dpath(physical), dpath(real))
-    var pj = path.resolve(real, 'package.json')
+    var pj = path.join(real, 'package.json')
     rpj(pj, function (er, pkg) {
       pkg = pkg || null
       var node
@@ -99,7 +99,7 @@ function loadChildren (node, cache, filterWith, cb) {
   debug('loadChildren', dpath(node.path))
   // don't let it be called more than once
   cb = once(cb)
-  var nm = path.resolve(node.path, 'node_modules')
+  var nm = path.join(node.path, 'node_modules')
   readdir(nm, function (er, kids) {
     // If there are no children, that's fine, just return
     if (er) return cb(null, node)
@@ -112,8 +112,8 @@ function loadChildren (node, cache, filterWith, cb) {
     if (l === 0) return cb(null, node)
 
     kids.forEach(function (kid) {
-      var kidPath = path.resolve(nm, kid)
-      var kidRealPath = path.resolve(node.realpath,'node_modules',kid)
+      var kidPath = path.join(nm, kid)
+      var kidRealPath = path.join(node.realpath,'node_modules',kid)
       loadNode(kidPath, kidRealPath, cache, then)
     })
 

--- a/rpt.js
+++ b/rpt.js
@@ -6,6 +6,23 @@ var once = require('once')
 var readdir = require('readdir-scoped-modules')
 var debug = require('debuglog')('rpt')
 
+function asyncForEach (items, todo, done) {
+  var remaining = items.length
+  if (remaining === 0) return done()
+  var seenErr
+  items.forEach(function (item) {
+    todo(item, handleComplete)
+  })
+  function handleComplete (err) {
+    if (seenErr) return
+    if (err) {
+      seenErr = true
+      return done(err)
+    }
+    if (--remaining === 0) done()
+  }
+}
+
 function dpath (p) {
   if (!p) return ''
   if (p.indexOf(process.cwd()) === 0) {
@@ -77,30 +94,40 @@ Link.prototype.realpath = ''
 
 function loadNode (logical, physical, cache, cb) {
   debug('loadNode', dpath(logical))
-  fs.realpath(physical, function (er, real) {
+  return fs.realpath(physical, thenReadPackageJson)
+
+  var realpath
+  function thenReadPackageJson (er, real) {
     if (er) return cb(er)
     debug('realpath l=%j p=%j real=%j', dpath(logical), dpath(physical), dpath(real))
     var pj = path.join(real, 'package.json')
-    rpj(pj, function (er, pkg) {
-      pkg = pkg || null
-      var node
-      if (physical === real) {
-        node = new Node(pkg, logical, physical, er, cache)
-      } else {
-        node = new Link(pkg, logical, physical, real, er, cache)
-      }
+    realpath = real
+    return rpj(pj, thenCreateNode)
+  }
+  function thenCreateNode (er, pkg) {
+    pkg = pkg || null
+    var node
+    if (physical === realpath) {
+      node = new Node(pkg, logical, physical, er, cache)
+    } else {
+      node = new Link(pkg, logical, physical, realpath, er, cache)
+    }
 
-      cb(null, node)
-    })
-  })
+    cb(null, node)
+  }
 }
 
 function loadChildren (node, cache, filterWith, cb) {
   debug('loadChildren', dpath(node.path))
-  // don't let it be called more than once
+  // needed 'cause we process all kids async-like and errors
+  // short circuit, so we have to be sure that after an error
+  // the cbs from other kids don't result in calling cb a second
+  // (or more) time.
   cb = once(cb)
   var nm = path.join(node.path, 'node_modules')
-  readdir(nm, function (er, kids) {
+  return readdir(nm, thenLoadKids)
+
+  function thenLoadKids (er, kids) {
     // If there are no children, that's fine, just return
     if (er) return cb(null, node)
 
@@ -108,26 +135,24 @@ function loadChildren (node, cache, filterWith, cb) {
       return kid[0] !== '.' && (!filterWith || filterWith(node, kid))
     })
 
-    var l = kids . length
-    if (l === 0) return cb(null, node)
-
-    kids.forEach(function (kid) {
-      var kidPath = path.join(nm, kid)
-      var kidRealPath = path.join(node.realpath,'node_modules',kid)
-      loadNode(kidPath, kidRealPath, cache, then)
-    })
-
-    function then (er, kid) {
-      if (er) return cb(er)
-
+    asyncForEach(kids, thenLoadNode, thenSortChildren)
+  }
+  function thenLoadNode (kid, done) {
+    var kidPath = path.join(nm, kid)
+    var kidRealPath = path.join(node.realpath,'node_modules',kid)
+    loadNode(kidPath, kidRealPath, cache, andAddNode(done))
+  }
+  function andAddNode (done) {
+    return function (er, kid) {
+      if (er) return done(er)
       node.children.push(kid)
-      kid.parent = node
-      if (--l === 0) {
-        sortChildren(node)
-        return cb(null, node)
-      }
+      done()
     }
-  })
+  }
+  function thenSortChildren (er) {
+    sortChildren(node)
+    cb(er, node)
+  }
 }
 
 function sortChildren (node) {
@@ -147,27 +172,25 @@ function loadTree (node, did, cache, filterWith, cb) {
 
   did[node.realpath] = true
 
+  // needed 'cause we process all kids async-like and errors
+  // short circuit, so we have to be sure that after an error
+  // the cbs from other kids don't result in calling cb a second
+  // (or more) time.
   cb = once(cb)
-  loadChildren(node, cache, filterWith, function (er, node) {
+  return loadChildren(node, cache, filterWith, thenProcessChildren)
+
+  function thenProcessChildren (er, node) {
     if (er) return cb(er)
 
     var kids = node.children.filter(function (kid) {
       return !did[kid.realpath]
     })
 
-    var l = kids.length
-    if (l === 0) return cb(null, node)
-
-    kids.forEach(function (kid, index) {
-      loadTree(kid, did, cache, filterWith, then)
-    })
-
-    function then (er, kid) {
-      if (er) return cb(er)
-
-      if (--l === 0) cb(null, node)
-    }
-  })
+    return asyncForEach(kids, loadTreeForKid, cb)
+  }
+  function loadTreeForKid (kid, done) {
+    loadTree(kid, did, cache, filterWith, done)
+  }
 }
 
 function rpt (root, filterWith, cb) {
@@ -175,16 +198,27 @@ function rpt (root, filterWith, cb) {
     cb = filterWith
     filterWith = null
   }
-  fs.realpath(root, function (er, realRoot) {
+  var cache = Object.create(null)
+  var topErr
+  var tree
+  return fs.realpath(root, thenLoadNode)
+
+  function thenLoadNode (er, realRoot) {
     if (er) return cb(er)
     debug('rpt', dpath(realRoot))
-    var cache = Object.create(null)
-    loadNode(root, realRoot, cache, function (er, node) {
-      // if there's an error, it's fine, as long as we got a node
-      if (!node) return cb(er)
-      loadTree(node, {}, cache, filterWith, function (lter, tree) {
-        cb(er && er.code !== 'ENOENT' ? er : lter, tree)
-      })
-    })
-  })
+    loadNode(root, realRoot, cache, thenLoadTree)
+  }
+  function thenLoadTree(er, node) {
+    // even if there's an error, it's fine, as long as we got a node
+    if (node) {
+      topErr = er
+      tree = node
+      loadTree(node, {}, cache, filterWith, thenHandleErrors)
+    } else {
+      cb(er)
+    }
+  }
+  function thenHandleErrors (er) {
+    cb(topErr && topErr.code !== 'ENOENT' ? topErr : er, tree)
+  }
 }

--- a/test/symlinked-node-modules.js
+++ b/test/symlinked-node-modules.js
@@ -1,0 +1,70 @@
+'use strict'
+var path = require('path')
+var test = require('tap').test
+var rpt = require('../rpt.js')
+var Tacks = require('tacks')
+var File = Tacks.File
+var Symlink = Tacks.Symlink
+var Dir = Tacks.Dir
+
+var workdir = path.join(__dirname, path.basename(__filename, '.js'))
+var fixture = new Tacks(Dir({
+  bar: Dir({
+    'package.json': File({
+      name: 'bar',
+      version: '1.0.0'
+    })
+  }),
+  'linked-node-modules': Dir({
+    bar: Symlink('../bar'),
+    foo: Dir({
+      'package.json': File({
+        name: 'foo',
+        version: '1.0.0'
+      })
+    })
+  }),
+  example: Dir({
+    node_modules: Symlink('../linked-node-modules/'),
+    'package.json': File({
+      name: 'example',
+      version: '1.0.0',
+    })
+  })
+}))
+
+function setup () {
+  cleanup()
+  fixture.create(workdir)
+}
+
+function cleanup () {
+  fixture.remove(workdir)
+}
+
+test('setup', function (t) {
+  setup()
+  t.done()
+})
+test('symlinked-node-modules', function (t) {
+  rpt(path.join(workdir, 'example'), function (err, tree) {
+    t.ifError(err)
+    t.is(tree.children.length, 2)
+    var childrenShouldBe = {
+      'foo': {isLink: false},
+      'bar': {isLink: true}
+    }
+    tree.children.forEach(function (child) {
+      var name = child.package.name
+      t.is(child.isLink, childrenShouldBe[name].isLink,
+        'is' + (childrenShouldBe[name].isLink ? '' : 'Not') + 'Link ' +
+        path.relative(workdir, child.path) + " + " +
+        path.relative(workdir, child.realpath))
+    })
+    t.done()
+  })
+})
+test('cleanup', function (t) {
+  cleanup()
+  t.done()
+})

